src/coreclr/gc/gccommon.cpp

```diff
@@ -91,26 +91,27 @@ struct changed_seg
 const int max_saved_changed_segs = 128;
 
 changed_seg saved_changed_segs[max_saved_changed_segs];
-int saved_changed_segs_count = 0;
+int saved_changed_segs_count = -1;
 
 void record_changed_seg (uint8_t* start, uint8_t* end,
                          size_t current_gc_index,
                          bgc_state current_bgc_state,
                          changed_seg_state changed_state)
 {
-    if (saved_changed_segs_count < max_saved_changed_segs)
-    {
-        saved_changed_segs[saved_changed_segs_count].start = start;
-        saved_changed_segs[saved_changed_segs_count].end = end;
-        saved_changed_segs[saved_changed_segs_count].gc_index = current_gc_index;
-        saved_changed_segs[saved_changed_segs_count].bgc = current_bgc_state;
-        saved_changed_segs[saved_changed_segs_count].changed = changed_state;
-        saved_changed_segs_count++;
-    }
-    else
-    {
-        saved_changed_segs_count = 0;
-    }
+#ifdef MULTIPLE_HEAPS
```

Comment:
I would make this `#if defined(MULTIPLE_HEAPS) && defined(USE_REGIONS)` and make `saved_changed_segs_count` a 64-bit integer.